common/rate-limiter/src/rate_limit.rs

```diff
@@ -0,0 +1,446 @@
+/// network requests, this may represent a request or a byte.  `Tokens` are the counters for the
+/// rate limiting, and when there are no `tokens` left in a `bucket`, the `key` is throttled.
+///
+/// ## Bucket
+/// A `bucket` is the tracker of the number of `tokens`.  It has a `bucket size`, and any additional
+/// tokens added to it will "spill" out of the `bucket`.  The `buckets` are filled at an `interval`
+/// with a given `fill rate`.
+///
+/// ## Interval
+/// The `interval` at which we refill *all* of the `buckets` in the token bucket filter. Configured
+/// across the whole token bucket filter.
+///
+/// ## Fill Rate
+/// The rate at which we fill a `bucket` with tokens. Configured per bucket.
+///
+/// ## Bucket Size
+/// Maximum size of a bucket.  A bucket saturates at this size.  Configured per bucket.
+///
+/// # Features
+/// ## Keys
+/// The token bucket takes any key as long as it's hashable.  This should allow it to apply to
+/// many applications that need rate limiters.
+///
+/// ## Bucket sizes and Rates
+/// ### Defaults
+/// There are defaults for bucket size and fill rate, which will apply to unknown keys.
+///
+/// ### Refill Interval
+/// Buckets are refilled automatically at an interval.  To do this synchronously, it calculates the
+/// number of intervals that have passed.  This is done synchronously and in the future may be done
+/// asynchronously.
+///
+pub struct TokenBucketRateLimiter<Key: Eq + Hash + Clone + Debug> {
+    buckets: RwLock<HashMap<Key, Arc<Mutex<Bucket>>>>,
+    default_bucket_size: NonZeroUsize,
+    default_fill_rate: NonZeroUsize,
+    open: bool,
+}
+
+impl<Key: Eq + Hash + Clone + Debug> TokenBucketRateLimiter<Key> {
+    pub fn new(default_bucket_size: NonZeroUsize, default_fill_rate: NonZeroUsize) -> Self {
+        assert!(
+            default_bucket_size >= default_fill_rate,
+            "Bucket size must be greater than or equal to fill rate"
+        );
+        Self {
+            buckets: RwLock::new(HashMap::new()),
+            default_bucket_size,
+            default_fill_rate,
+            open: false,
+        }
+    }
+
+    /// Used for testing and to not have a rate limiter
+    pub fn open() -> Self {
+        let max = NonZeroUsize::new(std::usize::MAX).unwrap();
+        Self {
+            buckets: RwLock::new(HashMap::new()),
+            default_bucket_size: max,
+            default_fill_rate: max,
+            open: true,
+        }
+    }
+
+    /// Retrieve bucket, or create a new one
+    pub fn bucket(&self, key: Key) -> Arc<Mutex<Bucket>> {
+        if !self.open {
+            self.bucket_inner(key, |size, rate| {
+                Arc::new(Mutex::new(Bucket::new(size, rate)))
+            })
+        } else {
+            self.bucket_inner(key, |_, _| Arc::new(Mutex::new(Bucket::open())))
+        }
+    }
+
+    fn bucket_inner<F: FnOnce(NonZeroUsize, NonZeroUsize) -> Arc<Mutex<Bucket>>>(
```

Comment:
Maybe typedef Arc<Mutex<Bucket>> to SharedBucket?

Reply:
I actually did that, and I guess it disappeared :/.  It'll make it cleaner